state/trackableDataTrie.go

```diff
@@ -144,129 +238,122 @@ func (tdaw *trackableDataTrie) SaveDirtyData(mainTrie common.Trie) ([]common.Tri
-
-func (tdaw *trackableDataTrie) updateTrieWithAutoBalancing(dtr dataTrie) ([]common.TrieData, error) {
-	oldValues := make([]common.TrieData, len(tdaw.dirtyData))
-
-	index := 0
-	for key, val := range tdaw.dirtyData {
-		oldEntry, err := tdaw.getOldKeyAndValWithCleanup(key)
-		if err != nil {
-			return nil, err
+func (tdaw *trackableDataTrie) getOldValue(key []byte, dataEntry dirtyData) core.TrieData {
+	if dataEntry.oldVersion.isValueKnown {
+		return core.TrieData{
+			Key:     key,
+			Value:   dataEntry.value,
+			Version: dataEntry.oldVersion.version,
 		}
+	}
 
-		oldValues[index] = oldEntry
-
-		err = tdaw.updateValInTrieWithAutoBalancing([]byte(key), val, dtr)
-		if err != nil {
-			return nil, err
+	if tdaw.enableEpochsHandler.IsAutoBalanceDataTriesEnabled() {
+		hashedKey := tdaw.hasher.Compute(string(key))
+		oldVal, _, err := tdaw.tr.Get(hashedKey)
+		if err == nil && len(oldVal) != 0 {
+			return core.TrieData{
+				Key:     hashedKey,
+				Value:   oldVal,
+				Version: core.AutoBalanceEnabled,
+			}
 		}
+	}
 
-		index++
+	oldVal, _, err := tdaw.tr.Get(key)
+	if err == nil && len(oldVal) != 0 {
+		return core.TrieData{
+			Key:     key,
+			Value:   oldVal,
+			Version: core.NotSpecified,
+		}
 	}
 
-	tdaw.dirtyData = make(map[string][]byte)
-	return oldValues, nil
+	newDataVersion := tdaw.getVersionForNewlyAddedData()
+	return core.TrieData{
+		Key:     tdaw.getKeyForVersion(key, newDataVersion),
+		Value:   nil,
+		Version: newDataVersion,
+	}
 }
 
-func (tdaw *trackableDataTrie) getOldKeyAndValWithCleanup(key string) (common.TrieData, error) {
-	hashedKey := tdaw.hasher.Compute(key)
+func (tdaw *trackableDataTrie) deleteOldEntryIfMigrated(key []byte, newData dirtyData, oldEntry core.TrieData) error {
+	if !tdaw.enableEpochsHandler.IsAutoBalanceDataTriesEnabled() {
+		return nil
+	}
 
-	oldVal, _, err := tdaw.tr.Get(hashedKey)
-	if err == nil && len(oldVal) != 0 {
-		return common.TrieData{
-			Key:     hashedKey,
-			Value:   oldVal,
-			Version: common.AutoBalanceEnabled,
-		}, nil
+	if oldEntry.Version == core.NotSpecified && newData.newVersion == core.AutoBalanceEnabled {
+		return tdaw.tr.Delete(key)
 	}
 
-	oldVal, _, err = tdaw.tr.Get([]byte(key))
+	return nil
+}
+
+func (tdaw *trackableDataTrie) modifyTrie(key []byte, dataEntry dirtyData, oldVal core.TrieData, dtr dataTrie) error {
+	shouldInsertInTrie, err := tdaw.deleteFromTrieIfNilVal(dataEntry.value, oldVal, key, dtr)
 	if err != nil {
-		return common.TrieData{}, err
+		return err
 	}
-
-	if len(oldVal) == 0 {
-		return common.TrieData{
-			Key:     hashedKey,
-			Value:   nil,
-			Version: common.NotSpecified,
-		}, nil
+	if !shouldInsertInTrie {
+		return nil
 	}
 
-	err = tdaw.tr.Delete([]byte(key))
+	newKey := tdaw.getKeyForVersion(key, dataEntry.newVersion)
+	value, err := tdaw.getValueForVersion(key, dataEntry.value, dataEntry.newVersion)
 	if err != nil {
-		return common.TrieData{}, err
+		return err
 	}
+	version := dataEntry.newVersion
```

Comment:
move this above line 329 and use version in the getKeyForVersion and getValueForVersion as well?

Reply:
Done.